tree/branch_node.go

```diff
@@ -0,0 +1,158 @@
+package tree
+
+import "fmt"
+
+// BranchNode represents a Node with an array of nodes and a sharedAddress
+// sharedAddress - part of the address that it links to
+// nodes are links to other nodes, which their addresses are suffixed by sharedAddress
+// [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, f ]
+//     [ x ] -> LeafNode
+//     [ not found ] -> EmptyNode
+//
+type BranchNode struct {
+	nodes         map[Unit]Node // TODO this can probably be an array if it help ?
+	sharedAddress []Unit
```

Comment:
`// sharedAddress is the shared prefix of all keys under this node`

Reply:
Much better description 👀